File: TeddyPet_BE/src/main/java/fpt/teddypet/application/dto/request/orders/order/OrderRequest.java
package fpt.teddypet.application.dto.request.orders.order;

import fpt.teddypet.domain.enums.payments.PaymentMethodEnum;
import jakarta.validation.Valid;
import jakarta.validation.constraints.*;
import java.util.List;

public record OrderRequest(
        @NotNull(message = "Phương thức thanh toán là bắt buộc") PaymentMethodEnum paymentMethod,

        // Optional: nếu có thì lấy từ địa chỉ đã lưu, không cần nhập thủ công
        Long userAddressId,

        // Các field này chỉ bắt buộc khi userAddressId = null hoặc guest checkout
        @Size(max = 100, message = "Tên người nhận tối đa 100 ký tự") String receiverName,

        @Pattern(regexp = "^(0|\\+84)[\\s.]?((3[2-9])|(5[689])|(7[06-9])|(8[1-689])|(9[0-46-9]))(\\d)[\\s.]?(\\d{3})[\\s.]?(\\d{3})$", message = "Số điện thoại không đúng định dạng Việt Nam") String receiverPhone,

        @Size(max = 500, message = "Địa chỉ quá dài (tối đa 500 ký tự)") String shippingAddress,

        @Size(max = 1000, message = "Ghi chú tối đa 1000 ký tự") String note,

        @NotEmpty(message = "Đơn hàng phải có ít nhất 1 sản phẩm") @Valid List<OrderItemRequest> items,

        @Size(max = 50, message = "Mã giảm giá tối đa 50 ký tự") String voucherCode,

        // Email cho guest checkout - bắt buộc khi không đăng nhập
        @Email(message = "Email không đúng định dạng") @Size(max = 255, message = "Email tối đa 255 ký tự") String guestEmail,

        // OTP Code cho guest xác thực email
        String otpCode,

        Double latitude,
        Double longitude

) {
}

File: teddypet_mobile/lib/data/models/request/order/order_request.dart
class OrderRequest {
  final String paymentMethod;
  final int? userAddressId;
  final String? receiverName;
  final String? receiverPhone;
  final String? shippingAddress;
  final String? note;
  final List<OrderItemRequest> items;
  final String? voucherCode;
  final double? latitude;
  final double? longitude;

  OrderRequest({
    required this.paymentMethod,
    this.userAddressId,
    this.receiverName,
    this.receiverPhone,
    this.shippingAddress,
    this.note,
    required this.items,
    this.voucherCode,
    this.latitude,
    this.longitude,
  });

  Map<String, dynamic> toJson() {
    return {
      'paymentMethod': paymentMethod, 
      if (userAddressId != null) 'userAddressId': userAddressId,
      if (receiverName != null) 'receiverName': receiverName,
      if (receiverPhone != null) 'receiverPhone': receiverPhone,
      if (shippingAddress != null) 'shippingAddress': shippingAddress,
      'note': (note == null || note!.isEmpty) ? null : note,
      'items': items.map((e) => e.toJson()).toList(),
      if (voucherCode != null) 'voucherCode': voucherCode,
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
    };
  }
}

class OrderItemRequest {
  final int variantId;
  final int quantity;

  OrderItemRequest({required this.variantId, required this.quantity});

  Map<String, dynamic> toJson() {
    return {
      'variantId': variantId,
      'quantity': quantity,
    };
  }
}
